File: app/src/main/java/baidu/com/testlibproject/ui/ConstrainLayoutActivity.java
package baidu.com.testlibproject.ui;

import android.app.Activity;
import android.content.Context;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.RequiresApi;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.Spinner;
import android.widget.TextView;

import baidu.com.testlibproject.R;

public class ConstrainLayoutActivity extends Activity {

    private static final String[] TEST_PROGRAM_ARRAY;

    static {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            TEST_PROGRAM_ARRAY = new String[]{
                    "相对位置demo",
                    "Margin demo",
                    "bias demo",
                    "ratio demo",
                    "chains demo",
                    "guideline demo",
                    "尺寸约束",
                    "RL性能测试",
                    "CL性能测试"
            };
        } else {
            TEST_PROGRAM_ARRAY = new String[]{
                    "相对位置demo",
                    "Margin demo",
                    "bias demo",
                    "ratio demo",
                    "尺寸约束",
                    "chains demo",
                    "guideline demo",
            };
        }
    }

    private Spinner mTestSpinner;
    private TextView mSpinnerTv;
    private FrameLayout mFrameLayout;

    private Context mContext;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_constrain_layout);
        mContext = this;

        mTestSpinner = findViewById(R.id.spinner);
        mSpinnerTv = findViewById(R.id.spinner_tv);
        mFrameLayout = findViewById(R.id.frame_layout);

        mSpinnerTv.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mTestSpinner.performClick();
            }
        });

        ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, TEST_PROGRAM_ARRAY);
        mTestSpinner.setAdapter(adapter);
        mTestSpinner.setSelection(0);
        mTestSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @RequiresApi(api = Build.VERSION_CODES.N)
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                mSpinnerTv.setText(TEST_PROGRAM_ARRAY[position]);
                mFrameLayout.removeAllViews();
                if (id == 0) {
                    addConstraintBasic();
                } else if (id == 1) {
                    addConstraintGoneMargin();
                } else if (id == 2) {
                    addConstraintBias();
                } else if (id == 3) {
                    addConstraintRatio();
                } else if (id == 4) {
                    addConstraintChains();
                } else if (id == 5) {
                    addConstraintGuideLine();
                } else if (id == 6) {
                    addConstraintDimension();
                } else if (id == 7) {
                    addRelativeLayoutTest();
                } else if (id == 8) {
                    addConstraintLayoutTest();
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
                mSpinnerTv.setText("");
            }
        });
    }

    private void addConstraintBasic() {
        View subView = LayoutInflater.from(mContext).inflate(R.layout.layout_constraint_basic, mFrameLayout, false);
        mFrameLayout.addView(subView);
    }

    private void addConstraintGoneMargin() {
        View subView = LayoutInflater.from(mContext).inflate(R.layout.layout_constraint_margin, mFrameLayout, false);
        Button normalMarginBtn = subView.findViewById(R.id.normal_margin_btn);
        final TextView normalTvB = subView.findViewById(R.id.normal_tv_b);
        normalMarginBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (normalTvB.getVisibility() == View.VISIBLE) {
                    normalTvB.setVisibility(View.GONE);
                } else {
                    normalTvB.setVisibility(View.VISIBLE);
                }
            }
        });

        Button goneMarginBtn = subView.findViewById(R.id.gone_margin_btn);
        final TextView goneTvB = subView.findViewById(R.id.gone_tv_b);
        goneMarginBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (goneTvB.getVisibility() == View.VISIBLE) {
                    goneTvB.setVisibility(View.GONE);
                } else {
                    goneTvB.setVisibility(View.VISIBLE);
                }
            }
        });
        mFrameLayout.addView(subView);
    }

    private void addConstraintBias() {
        View subView = LayoutInflater.from(mContext).inflate(R.layout.layout_constraint_bias, mFrameLayout, false);
        mFrameLayout.addView(subView);
    }

    private void addConstraintRatio() {
        View subView = LayoutInflater.from(mContext).inflate(R.layout.layout_constraint_ratio, mFrameLayout, false);
        mFrameLayout.addView(subView);
    }

    private void addConstraintChains() {
        View subView = LayoutInflater.from(mContext).inflate(R.layout.layout_constraint_chains, mFrameLayout, false);
        mFrameLayout.addView(subView);
    }

    private void addConstraintGuideLine() {
        View subView = LayoutInflater.from(mContext).inflate(R.layout.layout_constraint_guideline, mFrameLayout, false);
        mFrameLayout.addView(subView);
    }

    private void addConstraintDimension() {
        View subView = LayoutInflater.from(mContext).inflate(R.layout.layout_constraint_dimension, mFrameLayout, false);
        mFrameLayout.addView(subView);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    private void addRelativeLayoutTest() {
        ConstraintLayoutTestLayout testLayout = new ConstraintLayoutTestLayout(this);
        testLayout.testRelativeLayout();
        mFrameLayout.addView(testLayout);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    private void addConstraintLayoutTest() {
        ConstraintLayoutTestLayout testLayout = new ConstraintLayoutTestLayout(this);
        testLayout.testConstraintLayout();
        mFrameLayout.addView(testLayout);
    }
}

File: app/src/main/java/baidu/com/testlibproject/ui/ConstrainLayoutActivity.kt
package baidu.com.testlibproject.ui

import android.app.Activity
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.*
import android.widget.AdapterView.OnItemSelectedListener
import androidx.annotation.RequiresApi
import baidu.com.testlibproject.R

class ConstrainLayoutActivity : Activity() {

    companion object {
        private val TEST_PROGRAM_ARRAY: Array<String> = 
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            arrayOf(
                    "相对位置demo",
                    "Margin demo",
                    "bias demo",
                    "ratio demo",
                    "chains demo",
                    "guideline demo",
                    "尺寸约束",
                    "RL性能测试",
                    "CL性能测试"
            )
        } else {
            arrayOf(
                    "相对位置demo",
                    "Margin demo",
                    "bias demo",
                    "ratio demo",
                    "尺寸约束",
                    "chains demo",
                    "guideline demo")
        }
    }

    private lateinit var spinner: Spinner
    private lateinit var spinnerTv: TextView
    private lateinit var frameLayout: FrameLayout

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_constrain_layout)
        spinnerTv = findViewById(R.id.spinner_tv)
        spinner = findViewById(R.id.spinner)
        frameLayout = findViewById(R.id.frame_layout)

        spinnerTv.setOnClickListener { spinner.performClick() }
        val adapter = ArrayAdapter(this,
                android.R.layout.simple_list_item_1, TEST_PROGRAM_ARRAY)
        spinner.adapter = adapter
        spinner.setSelection(0)
        spinner.onItemSelectedListener = object : OnItemSelectedListener {
            @RequiresApi(api = Build.VERSION_CODES.N)
            override fun onItemSelected(parent: AdapterView<*>?, view: View, position: Int, id: Long) {
                spinnerTv.text = TEST_PROGRAM_ARRAY[position]
                frameLayout.removeAllViews()
                when (id) {
                    0L -> {
                        addConstraintBasic()
                    }
                    1L -> {
                        addConstraintGoneMargin()
                    }
                    2L -> {
                        addConstraintBias()
                    }
                    3L -> {
                        addConstraintRatio()
                    }
                    4L -> {
                        addConstraintChains()
                    }
                    5L -> {
                        addConstraintGuideLine()
                    }
                    6L -> {
                        addConstraintDimension()
                    }
                    7L -> {
                        addRelativeLayoutTest()
                    }
                    8L -> {
                        addConstraintLayoutTest()
                    }
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {
                spinnerTv.text = ""
            }
        }
    }

    private fun addConstraintBasic() {
        LayoutInflater.from(this).inflate(
            R.layout.layout_constraint_basic, frameLayout, true)
    }

    private fun addConstraintGoneMargin() {
        val subView = LayoutInflater.from(this).inflate(R.layout.layout_constraint_margin,
                frameLayout, false)
        val normalMarginBtn = subView.findViewById<Button>(R.id.normal_margin_btn)
        val normalTvB = subView.findViewById<TextView>(R.id.normal_tv_b)
        normalMarginBtn.setOnClickListener {
            if (normalTvB.visibility == View.VISIBLE) {
                normalTvB.visibility = View.GONE
            } else {
                normalTvB.visibility = View.VISIBLE
            }
        }
        val goneMarginBtn = subView.findViewById<Button>(R.id.gone_margin_btn)
        val goneTvB = subView.findViewById<TextView>(R.id.gone_tv_b)
        goneMarginBtn.setOnClickListener {
            if (goneTvB.visibility == View.VISIBLE) {
                goneTvB.visibility = View.GONE
            } else {
                goneTvB.visibility = View.VISIBLE
            }
        }
        frameLayout.addView(subView)
    }

    private fun addConstraintBias() {
        LayoutInflater.from(this).inflate(
            R.layout.layout_constraint_bias, frameLayout, true)
    }

    private fun addConstraintRatio() {
        LayoutInflater.from(this).inflate(
            R.layout.layout_constraint_ratio, frameLayout, true)
    }

    private fun addConstraintChains() {
        LayoutInflater.from(this).inflate(
            R.layout.layout_constraint_chains, frameLayout, true)
    }

    private fun addConstraintGuideLine() {
        LayoutInflater.from(this).inflate(
            R.layout.layout_constraint_guideline, frameLayout, true)
    }

    private fun addConstraintDimension() {
        LayoutInflater.from(this).inflate(R.layout.layout_constraint_dimension,
                frameLayout, true)
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    private fun addRelativeLayoutTest() {
        val testLayout = ConstraintLayoutTestLayout(this)
        testLayout.testRelativeLayout()
        frameLayout.addView(testLayout)
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    private fun addConstraintLayoutTest() {
        val testLayout = ConstraintLayoutTestLayout(this)
        testLayout.testConstraintLayout()
        frameLayout.addView(testLayout)
    }
}